Statistics/milestone 2

urweakness opened the pull request:
## README
Стек: SwiftUI + MVVM
Декомпозиция на доске: [README](https://github.com/users/pavelseleznev/projects/1/views/11)
Декомпозиция эпика: [table teamwork (view)](https://github.com/users/pavelseleznev/projects/1/views/7)
Декомпозиция эпика в md:  [DecompositionStatistics.md](https://github.com/user-attachments/files/24279217/DecompositionStatistics.md)
Эпик: [Statistics - table teamwork](https://github.com/pavelseleznev/iOS-FakeNFT-StarterProject-Extended-Public/issues/86)

Намеченное и фактическое время представлено в файле, а так же можно посмотреть тут https://github.com/users/pavelseleznev/projects/1/views/3?sliceBy%5Bvalue%5D=urweakness&filterQuery=-label%3ADecompositionFile+assignee%3Aurweakness+milestone%3A%22Milestone+2%22

## Что сделано
- Добавил логику на каждый кейс сортировки
- Заменил моки на реальные данные
- Дописал NFTService, NFTStorage для кеширования данных пользователя
- Добавил пагинацию на загрузку списка пользователей
- Заменил Kingfisher на AsyncImage
- Реализовал AsyncNFTs объект для потоковой загрузки NFT с анимациями и ненавязчивым лоадером справа наверху в тулбаре 
- Для AsyncNFTs добавил поддержку состояния загрузки для возможности отмены при выходе с экрана
- Сделал ошибку отмены загрузки как допустимую 
- Связал кэшированные данные с сетевыми, определяя кэшированные как истину перед тем как отправить запрос (при этом кеш загружается при старте приложения и обновляется по мере пользования)

## Чеклист
- [ + ] - Код собирается и запускается в Xcode
- [ + ] - Нет ворнингов
- [ + ] - Поддержка превью в разных стейтах (если UI)

## Ссылка на задачу/issue
- https://github.com/pavelseleznev/iOS-FakeNFT-StarterProject-Extended-Public/issues/86

urweakness committed:
[Feature] NFTStorage
- extended & upgraded

urweakness committed:
[Refactor] ObservedNetworkClient
- dummy removed
- discardables added
- getUsers method pagination added

urweakness committed:
[Feature] UsersRequests
- pagination + sorting added

urweakness committed:
[Feature] NFTModelContainer
- added to contain cart and favourite flags

urweakness committed:
[Refactor] NFTResponse
- Identifiable now
- mocks added

urweakness committed:
[Refactor] UserListItemResponse
- Hashable now

urweakness committed:
[Feature] AsyncNFTs
- added as async streaming api

urweakness committed:
[Feature] ObservedNetworkClient + UserRequests
- sortOption added

urweakness committed:
[Feature] CoordinatorView + Coordinator
- on app launch profile loading added
- alert on error added with retry

urweakness committed:
[Refactor] UserListCell
- restrictions fix
- AsyncImage usage instead of KF

urweakness committed:
[Feature] StatisitcsNFTCollectionVIew
- no VM needed, removing
- view contains service, nftsIDs & loadingState

urweakness committed:
[Refactor] StatisticsProfileView
- viewModel data usage instead of mock

urweakness committed:
[Refactor] StatisticsView
- little fixes

urweakness committed:
[Refactor] StatisticsProfileViewModel
- model nftsIDs usage instead of mocks

urweakness committed:
[Refactor] StatisticsViewModel
- real data usage instead of mocks

urweakness committed:
[Refactor] Dialogs
- simplified sort usage

urweakness committed:
[Refactor] NFTCollectionView
- asyncNFTs support added
- loading error added with failed nfts retry
- preview mocks changed

urweakness committed:
[Refactor] NFTCartCellView
- NFTResponse instead of NFTModel usage
- favourites passed throw

urweakness committed:
[Refactor] NFTCompactCellView
- NFTResponse instead of NFTModel usage
- favourites passed throw

urweakness committed:
[Refactor] NFTMyCellView
- NFTResponse instead of NFTModel usage
- favourites passed throw

urweakness committed:
[Refactor] NFTCartCellView
- NFTModelContainer instead of NFTModel usage
- cost appearance fix
- cart state appearance added

urweakness committed:
[Refactor] NFTCostView
- NFTResponse instead of NFTModel usage

urweakness committed:
[Refactor] NFTImageView
- AsyncImage instead of KF
- favorited usage

urweakness committed:
[Refactor] NFTNameRateAuthorView
- NFTReponse instead of NFTModel
- from field appearance fix

urweakness committed:
[Refactor] ProfileHeader
- AsyncImage instead of KF

urweakness committed:
[Refactor] RatingPreview
- rating type support changed

urweakness committed:
[Refactor] StatisticsSortActionsViewModifier
- SortOption parameter variable added
- activeSortOption is binding now, viewModel contains orig

urweakness committed:
[Refactor] AppContainer
- rename

urweakness committed:
[Refactor] Page
- case assoc type change

urweakness committed:
[Feature] NFTService
- protocol updated
- service updated

urweakness committed:
[Feature] NFTCollectionView
- animations added

urweakness committed:
[Refactor] NFTImageVIew
- loading background added
- safe url unwrapping

urweakness committed:
[Refactor] DecompositionStatistics.md edited

urweakness committed:
[Feature] DataLoader
- cancellation error skipping

urweakness committed:
[Feature] AsyncNFTs
- fetching task managment
- nft load timout added

urweakness committed:
[Feature] NFTCollectionView
- onDissapear handling added

numpadCrusader reviewed (approved):
Отличная работа, Тимур! Все чисто и структурно. С моей стороны нечего добавить

pavelseleznev reviewed (approved):
С моей стороны тоже ок.

Kingfisher можно удалит, так как не используется.

Alex-rev-dev commented on iOS-FakeNFT-Extended/Foundation/MemoryStorage/NftStorage.swift:
```diff
@@ -1,19 +1,61 @@
+	private var cart = Set<String>()
+}
+
+// --- cart ---
```
Предлагаю вот такие комментарии:
// MARK: - cart

urweakness committed:
[Codestyle] NFTStorage
- comments updated

Alex-rev-dev commented on iOS-FakeNFT-Extended/Modules/Coordinator/View/CoordinatorView.swift:
```diff
@@ -28,15 +27,27 @@ struct CoordinatorView: View {
 				.allowsHitTesting(coordinator.appContainer.api.loadingState != .fetching)
+				.applyRepeatableAlert(
+					isPresneted: .constant(coordinator.appContainer.api.loadingState == .error),
+					message: "Не удалось получить данные",
```
предлагаю вынести значение в константы

urweakness replied:
предлагаю вынести это в структуру с константами имен алертов для всего приложения, когда все эпики будут готовы
сейчас не вижу смысла плодить доп сущности, предлагаю добавить TODO

urweakness committed:
[Feature] TODO's added

Alex-rev-dev commented:
У тебя получился очень качественный структурированный код, так держать!

urweakness committed:
[Feature] UserList sort option stored now

urweakness committed:
[Feature] AsyncNFTs
- base nft sort by id added

urweakness committed:
[Feature] LoadingShimmerPlaceholderView

urweakness committed:
[Feature] Loading shimmer support added to subviews

urweakness committed:
[Feature] AsyncNFTs
- now supports key-value nfts storing, allows saving order to show placeholder when nft is nil
